app/publish/publish.go

```diff
@@ -0,0 +1,132 @@
+package publish
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+	"strconv"
+
+	"github.com/lbryio/lbrytv/app/proxy"
+	"github.com/lbryio/lbrytv/app/users"
+	"github.com/lbryio/lbrytv/internal/lbrynet"
+
+	ljsonrpc "github.com/lbryio/lbry.go/extras/jsonrpc"
+	"github.com/ybbus/jsonrpc"
+)
+
+const uploadPath = "/tmp"
+
+const fileField = "file"
+const jsonrpcPayloadField = "json_payload"
+
+type Publisher interface {
+	Publish(string, string, []byte) ([]byte, error)
+}
+
+type LbrynetPublisher struct{}
+
+type uploadHandler struct {
+	Publisher  Publisher
+	uploadPath string
+}
+
+func NewUploadHandler(uploadPath string, publisher Publisher) uploadHandler {
+	return uploadHandler{
+		Publisher:  publisher,
+		uploadPath: uploadPath,
+	}
+}
+
+func (p *LbrynetPublisher) Publish(filePath, accountID string, rawQuery []byte) ([]byte, error) {
+	// var rpcParams *lbrynet.PublishParams
+	// var rpcParams *ljsonrpc.StreamCreateOptions
+	rpcParams := struct {
+		Name                          string  `json:"name"`
+		Bid                           string  `json:"bid"`
+		FilePath                      string  `json:"file_path,omitempty"`
+		FileSize                      *string `json:"file_size,omitempty"`
+		IncludeProtoBuf               bool    `json:"include_protobuf"`
+		Blocking                      bool    `json:"blocking"`
+		*ljsonrpc.StreamCreateOptions `json:",flatten"`
+	}{}
+
+	query, err := proxy.NewQuery(rawQuery)
+	if err != nil {
+		panic(err)
+	}
+
+	if err := query.ParamsToStruct(&rpcParams); err != nil {
+		panic(err)
+	}
+
+	if rpcParams.FilePath != "__POST_FILE__" {
+		panic("unknown file_path content")
+	}
+
+	bid, err := strconv.ParseFloat(rpcParams.Bid, 64)
+	rpcParams.FilePath = filePath
+	rpcParams.AccountID = &accountID
+
+	result, err := lbrynet.Client.StreamCreate(rpcParams.Name, filePath, bid, *rpcParams.StreamCreateOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	rpcResponse := jsonrpc.RPCResponse{Result: result}
+	serialized, err := json.MarshalIndent(rpcResponse, "", "  ")
+	if err != nil {
+		return nil, err
+	}
+	return serialized, nil
+}
+
+func (h uploadHandler) Handle(w http.ResponseWriter, r *users.AuthenticatedRequest) {
+	if !r.IsAuthenticated() {
+		var authErr Error
+		if r.AuthFailed() {
+			authErr = NewAuthError(r.AuthError)
+		} else {
+			authErr = ErrUnauthorized
+		}
+		w.WriteHeader(http.StatusOK)
+		w.Write(authErr.AsBytes())
+		return
+	}
+	file, header, err := r.FormFile("file")
+	if err != nil {
+		panic(err)
+	}
+	defer file.Close()
```

Comment:
this is also ignoring an error. There is also a good article about why not to do this. https://www.joeshaw.org/dont-defer-close-on-writable-files/

I handle this in our go apps like this https://github.com/lbryio/internal-apis/blob/master/app/db/init.go#L167-L183

Its just a couple examples from one file but calling the wrapper with defer is better so the error is not ignored.

Reply:
That's a very good point, thanks for bringing this to my attention